File: src/main/java/de/spries/fleetcommander/model/core/universe/UniverseFactory.java
package de.spries.fleetcommander.model.core.universe;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import de.spries.fleetcommander.model.core.Player;

public class UniverseFactory {
	public static final int PLANET_COUNT = 100;
	private static final int MIN_PLANET_OFFSET = -2;
	private static final int MAX_PLANET_OFFSET = 2;
	private static final Random rand = new Random();

	/**
	 * Prototype implementation for testing purposes
	 */
	public static Universe generate(List<Player> players) {
		List<Planet> planets = new ArrayList<>(PLANET_COUNT);

		for (int row = 0; row < 10; row++) {
			for (int col = 0; col < 10; col++) {
				int offsetX = randomOffset();
				int offsetY = randomOffset();

				Planet planet = new Planet(col * 10 + 5 + offsetX, row * 10 + 5 + offsetY);
				planets.add(planet);
			}
		}

		Collections.shuffle(planets);

		planets = planets.subList(0, planets.size() / (6 - players.size()));

		int planetId = 0;
		for (Planet planet : planets) {
			planet.setId(planetId++);
		}

		for (int i = 0; i < players.size(); i++) {
			Planet oldUninhabitedPlanet = planets.get(i);
			Planet newHomePlanet = new Planet(oldUninhabitedPlanet.getX(), oldUninhabitedPlanet.getY(), players.get(i));
			newHomePlanet.setId(oldUninhabitedPlanet.getId());
			planets.set(i, newHomePlanet);
		}

		return new Universe(planets);
	}

	private static int randomOffset() {
		return rand.nextInt(MAX_PLANET_OFFSET - MIN_PLANET_OFFSET + 1) + MIN_PLANET_OFFSET;
	}

}

File: src/main/java/de/spries/fleetcommander/model/core/universe/UniverseFactory.kt
package de.spries.fleetcommander.model.core.universe

import de.spries.fleetcommander.model.core.Game
import de.spries.fleetcommander.model.core.Player
import java.util.ArrayList
import java.util.Random

object UniverseFactory {
    private const val PLANET_COUNT = 100
    private const val MIN_PLANET_OFFSET = -2
    private const val MAX_PLANET_OFFSET = 2
    private val rand = Random()

    /**
     * Prototype implementation for testing purposes
     */
    fun generate(players: Collection<Player>): Universe {
        var planets: MutableList<Planet> = ArrayList(PLANET_COUNT)

        for (row in 0..9) {
            for (col in 0..9) {
                val offsetX = randomOffset()
                val offsetY = randomOffset()

                val planet = Planet(col * 10 + 5 + offsetX, row * 10 + 5 + offsetY)
                planets.add(planet)
            }
        }

        planets.shuffle()

        planets = planets.subList(0, planets.size / (1 + Game.MAX_PLAYERS - players.size))

        var planetId = 0
        for (planet in planets) {
            planet.id = planetId++
        }

        for ((i, player) in players.withIndex()) {
            val oldUninhabitedPlanet = planets[i]
            val newHomePlanet = Planet(oldUninhabitedPlanet.x, oldUninhabitedPlanet.y, player)
            newHomePlanet.id = oldUninhabitedPlanet.id
            planets[i] = newHomePlanet
        }

        return Universe(planets)
    }

    private fun randomOffset(): Int {
        return rand.nextInt(MAX_PLANET_OFFSET - MIN_PLANET_OFFSET + 1) + MIN_PLANET_OFFSET
    }

}